test_common/harness/kernelHelpers.cpp

```diff
@@ -809,6 +810,71 @@ int create_single_kernel_helper(cl_context context,
+        // version supported by the device. This allows calling code to force a
+        // particular CL C version if it is required, but also means that
+        // callers need not specify a version if they want to assume the most
+        // recent CL C.
+
+        // First get all the devices in the context and find the minumum
+        // supported OpenCL C version.
+        size_t devices_size_in_bytes{};
+        auto error = clGetContextInfo(context, CL_CONTEXT_DEVICES, 0, nullptr,
+                                      &devices_size_in_bytes);
+        test_error(error, "clGetDeviceInfo failed for CL_CONTEXT_DEVICES");
+        std::vector<cl_device_id> devices(devices_size_in_bytes
+                                          / sizeof(cl_device_id));
+        error =
+            clGetContextInfo(context, CL_CONTEXT_DEVICES, devices_size_in_bytes,
+                             devices.data(), nullptr);
+        auto current_version = get_device_cl_version(devices[0]);
+        std::for_each(std::next(devices.begin()), devices.end(),
+                      [&current_version](cl_device_id device) {
+                          auto device_version = get_device_cl_version(device);
+                          // OpenCL 3.0 is not backwards compatible with 2.0.
+                          // If we have 3.0 and 2.0 in the same driver we
+                          // use 1.2.
+                          if (((device_version >= Version(2, 0)
+                                && device_version < Version(3, 0))
+                               && current_version >= Version(3, 0))
+                              || (device_version >= Version(3, 0)
+                                  && (current_version >= Version(2, 0)
+                                      && current_version < Version(3, 0))))
+                          {
+                              current_version = Version(1, 2);
+                          }
+                          else
+                          {
+                              current_version =
+                                  (std::min)(device_version, current_version);
+                          }
+                      });
+
+        std::string cl_std{};
+        if (current_version >= Version(3, 0))
+        {
+            cl_std = "-cl-std=CL3.0";
+        }
+        else if (current_version >= Version(2, 0)
+                 && current_version < Version(3, 0))
+        {
+            cl_std = "-cl-std=CL2.0";
+        }
+        else
+        {
+            // If the -cl-std build option is not specified, the highest OpenCL
+            // C 1.x language version supported by each device is used when
+            // compiling the program for each device.
+            cl_std = "";
+        }
+        std::string updated_build_options = buildOptions ? buildOptions : "";
+        updated_build_options += cl_std;
+        buildOptions = updated_build_options.c_str();
```

Comment:
This will end up pointing at memory that has been freed.

Reply:
I've updated this to use a `std::string` declared at the top of the function, it's not ideal but I guess this is a results of trying to use c-style strings and `std::string` together.